### DISCUSSION TITLE
Do we need to calculate Schmid Factor (SF) prior to calculating mPrime?

### DISCUSSION THREAD

**Discussion by Mustafa-Rifat** [mtex-toolbox/mtex#1414] [Ask Anything] [answered]
Hello, I am interested in determining the slip transfer ability through grain boundaries. I found out that calculation of mPrime [as in this page](https://mtex-toolbox.github.io/HomepageOld/files/doc/SlipTransmition.html) under compatibility of slip systems is a way to do that. However, as per my understanding, there is no need to calculate Schmid Factor to get the m' plot. My questions are:
1. Is my understanding correct about no need of calculating the SF?
2. In the [demo page](https://mtex-toolbox.github.io/HomepageOld/files/doc/SlipTransmition.html) it is written "Next, we want to analyze, how much geometrically compatible the slip systems with highest Schmid factor are at the grain boundaries". I am curious how mPrime function knows which slip systems have highest Schmid factor. According to [this page](https://mtex-toolbox.github.io/slipSystem.mPrime.html) explaining the function, it just takes two slip systems corresponding to two neighboring grains as input. Without even calculating SF, how does it know slip systems with highest SF?
3. If I am particularly interested to know about the mPrime value in the GB between two grains, rather than a map, how can I do that?
My current code (after importing data) is given below:
``` matlab
ebsd = loadEBSD(fname,CS,'interface','ctf','convertEuler2SpatialReferenceFrame');
ipfKey = ipfHSVKey(ebsd('Ni-superalloy'));
ebsd=ebsd('Ni-superalloy');
[grains,ebsd.grainId] = calcGrains(ebsd,'angle',15*degree);
grains_small_angle = calcGrains(ebsd,'angle',1*degree);
plot(ebsd,ebsd.orientations,'micronbar','off')
hold on
plot(grains.boundary,'linewidth',2)
hold on
plot(grains_small_angle.boundary,'linewidth',1)
hold off
sS = slipSystem.fcc(ebsd.CS)
sS = sS.symmetrise;
sSLocal = grains.meanOrientation * sS;
gB = grains.boundary('indexed');
id = gB.grainId;
 mP = mPrime(sSLocal(id(:,1)),sSLocal(id(:,2)));
hold on
plot(gB,mP,'linewidth',3)
mtexColorbar
```

The figure I am getting is like this:
![image](https://user-images.githubusercontent.com/60717121/165802008-1ef83328-84ef-4b8c-ab6f-bdb0aebd0abf.png)

**Comment by kilir** [comment, marked as answer]
Hi,
with respect to (1), yes, you are correct.
Wrt (2), it's referrign to the quiver plot on top, so one could look at those boundaries are a) being permissive to slip and b) having adjacent grains with slip systems doing the save / or different jobs. That could be more clearly stated there, suggestions welcome.
Wrt (3) you probably do:
```
gB = grains('id',[x,y]).boundary('indexed');
id = gB.grainId;
```
where x and y are the ids of your grain of interest. Is that what you mean?

Cheers,
Rüdiger

**Reply by Mustafa-Rifat** [reply]
Thanks a lot. I could solve problem stated in question #3. 
However, I am afraid I am still not clear about how the mPrime function works (apologize for that). Does it take all the slip planes and slip directions (from the 12 slip systems in fcc) in consideration to calculate `m′=(n⃗ in⋅n⃗ out)(d⃗ in⋅d⃗ out)`?

**Reply by filippeof** [reply]
Hi @Mustafa-Rifat,
The m' calculation in mtex should take one slip system for each grain only. To get all possible combinations of potential slip systems for the "in" and "out" grains you would probably need something like this:
```matlab
sS = slipSystem.fcc(ebsd.CS);
sS = sS.symmetrise;
sSLocal = grains.meanOrientation * sS;
gB = grains.boundary('indexed');
id = gB.grainId;
sz=size(sSLocal);
mp=nan([sz,sz(2)]);
for ii=1:size(sSLocal,2)
    for jj=1:size(sSLocal,2)
        mP(:,ii,jj) = mPrime(sSLocal(id(:,1),ii),sSLocal(id(:,2),jj));
    end
end

```
You could simplify things by assuming a 'most-likely' slip system for each grain. For example, the slip system that has the highest Schmid's factor (as in https://mtex-toolbox.github.io/HomepageOld/files/doc/SlipTransmition.html)

Let me know if you have questions on how to implement it.

ps. There's a great explanation of the m' factor calculations here:
https://stabix.readthedocs.io/en/latest/

Best,
Filippe

**Reply by Mustafa-Rifat** [reply]
Thanks Filippe! 
I saw the page in github and that was exactly what I had confusion about. Have a look at this image:
![image](https://user-images.githubusercontent.com/60717121/166074228-c94bf3e7-9f2d-4254-a2e6-bd2e4378df79.png)
It calculates m' using this line:
`mP = mPrime(sSGrain(id(:,1)),sSGrain(id(:,2)));`
As you said, to calculate the m', it is taking one slip system for each grain only. What I was wondering is how it chooses the slip system? Does it randomly chooses one? 
We do not need to calculate SF to get the slip transmission plot generated here. Let's say, we do not calculate SF before (because I do not have information regarding the stress tensor near grain boundaries in my loading condition) and directly move to this plot. What will the mPrime code do then?

**Reply by ralfHielscher** [reply]
Hi Mustafa,

the sections within one documentation page are build on top of each other. In this case the variable `sSGrain` was computed in the previous section an contains the slip system with the maximum Schmidfactor for each grain. This is then used as input to `mPrime` in the section you are quoting.

Ralf.

**Reply by Mustafa-Rifat** [reply]
Thanks a lot! This explains it!

**Reply by Mustafa-Rifat** [reply]
I have one follow up question. As recommended by Ralf, now I am calculating stress tensor from a FEA model that I have and using that to calculate Schmid Factor (SF). Now the stress tensor that I get from FEA, _**do we need to transfer it to crystal coordinate to calculate SF**_? The SF documentation page states in the beginning "So far we have always assumed that the stress tensor is already given relatively to the crystal coordinate system. Next, we want to examine the case where the stress is given in specimen coordinates and we know the orientation of the crystal." I believe the FEA data that is for specimen coordinate (it is a regular FEA model, not CPFEM).

My current code is like below:
```
CS = {... 
  crystalSymmetry('m-3m', [3.57 3.57 3.57], 'mineral', 'Ni', 'color', 'light blue'),...
  crystalSymmetry('m-3m', [3.57 3.57 3.57], 'mineral', 'Ni-superalloy', 'color', 'light blue')};

% plotting convention
setMTEXpref('xAxisDirection','east');
setMTEXpref('zAxisDirection','outOfPlane');

%% Specify File Names

% path to files
pname = 'C:\Matlab Directory';
 
% which files to be imported
fname = [pname '\Sample 2 postscratch Specimen 1 Site 1 Map Data 4.ctf'];

%% Import the Data

% create an EBSD variable containing the data
ebsd = loadEBSD(fname,CS,'interface','ctf',...
  'convertEuler2SpatialReferenceFrame');
%ebsd = rotate(ebsd,rotation('axis',yvector,'angle',180*degree));
ipfKey = ipfHSVKey(ebsd('Ni-superalloy'));
ebsd=ebsd('Ni-superalloy');
%grains = smooth(grains,5);
[grains,ebsd.grainId] = calcGrains(ebsd,'angle',15*degree);
grains_small_angle = calcGrains(ebsd,'angle',1*degree);
plot(ebsd,ebsd.orientations,'micronbar','off')
hold on
plot(grains.boundary,'linewidth',2)
hold on
plot(grains_small_angle.boundary,'linewidth',1)
hold off
sS = slipSystem.fcc(ebsd.CS)
sS = sS.symmetrise;
sSLocal = grains.meanOrientation * sS;
%This is the stress tensor I get from FEA
**stress_FEA=[-8.20594E08 2.63848E07 1.79507E07;2.63848E07 -6.93397E08 1.70987E07;1.79507E07 1.70987E07 -1.04309E08];
sigma = stressTensor(stress_FEA);
SF = sSLocal.SchmidFactor(sigma);**
[SFMax,active] = max(SF,[],2);
figure
plot(grains,SFMax,'micronbar','off','linewidth',2)
mtexColorbar location southoutside
sSactive = grains.meanOrientation .* sS(active);
hold on
% visualize the trace of the slip plane
quiver(grains,sSactive.trace,'color','b')

% and the slip direction
quiver(grains,sSactive.b,'color','r')
hold off

sSGrain = grains.meanOrientation .* sS(active);
figure
plot(grains,'FaceColor',0.8*[1 1 1],'figSize','large')
gB = grains.boundary('indexed');
id = gB.grainId;
mP = mPrime(sSGrain(id(:,1)),sSGrain(id(:,2)));
hold on
plot(gB,mP,'linewidth',3)
hold on
plot(grains_small_angle.boundary,'linewidth',1)
mtexColorbar
```
